api/routes/image_pack.go

```diff
@@ -0,0 +1,197 @@
+	api "github.com/uncharted-distil/distil/api/model"
+	"github.com/uncharted-distil/distil/api/util"
+)
+
+// ImagePackRequest is the expected request format for the route (Band can be an empty string)
+type ImagePackRequest struct {
+	Dataset  string   `json:"dataset"`
+	ImageIDs []string `json:"imageIds"`
+	Band     string   `json:"band,omitempty"`
+}
+
+// ImagePackResult is the expected post result for MultiBandImagePackHandler
+type ImagePackResult struct {
+	ImagesBuffer [][]byte `json:"images"`
+	ImageIDs     []string `json:"imageIds"`
+}
+type chanStruct struct {
+	data [][]byte
+	IDs  []string
+}
+
+func postParamsToImagePackRequest(r *http.Request) (*ImagePackRequest, error) {
+	body, err := ioutil.ReadAll(r.Body)
+	defer r.Body.Close()
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse POST request")
+	}
+	result := &ImagePackRequest{}
+	err = json.Unmarshal(body, result)
+	return result, err
+}
+
+// MultiBandImagePackHandler fetches individual band images and combines them into a single RGB image using the supplied mapping.
+func MultiBandImagePackHandler(ctor api.MetadataStorageCtor, dataCtor api.DataStorageCtor, config env.Config) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		params, err := postParamsToImagePackRequest(r)
+		if err != nil {
+			handleError(w, err)
+			return
+		}
+		// default to getImages
+		funcPointer := getImages
+		if params.Band != "" {
+			// if band is not empty then get multiBandImages
+			funcPointer = getMultiBandImages
+		}
+		// channel for threads to communicate
+		result := make(chan chanStruct)
+		// ImageThreadPool is an environment variable defaults to 2 (works great with 6)
+		for i := 0; i < config.ImageThreadPool; i++ {
+			go funcPointer(params, i, config.ImageThreadPool, result, ctor, dataCtor)
+		}
+		imagesBuffer := [][]byte{}
+		IDs := []string{}
+		for i := 0; i < config.ImageThreadPool; i++ {
+			// no guaruntee of threads finishing in order so we supply the IDs back as well
+			r := <-result
+			imagesBuffer = append(imagesBuffer, r.data...)
+			IDs = append(IDs, r.IDs...)
+		}
+		err = handleJSON(w, ImagePackResult{
+			ImagesBuffer: imagesBuffer,
+			ImageIDs:     IDs,
+		})
+		if err != nil {
+			handleError(w, errors.Wrap(err, "unable marshal dataset result into JSON"))
+			return
+		}
+	}
+}
+func getImages(imagePackRequest *ImagePackRequest, threadID int, numThreads int, result chan chanStruct, ctor api.MetadataStorageCtor, dataCtor api.DataStorageCtor) {
+	temp := [][]byte{}
+	IDs := []string{}
+	for i := threadID; i < len(imagePackRequest.ImageIDs); i += numThreads {
+		imageID := imagePackRequest.ImageIDs[i]
+		// get metadata client
+		storage, err := ctor()
+		if err != nil {
+			continue
+		}
+
+		res, err := storage.FetchDataset(imagePackRequest.Dataset, false, false, false)
+		if err != nil {
+			continue
+		}
+
+		sourcePath := env.ResolvePath(res.Source, res.Folder)
```

Comment:
the storage, fetch and source path are common across all images and should be done outside of the loop